### ISSUE TITLE
Selection types should not expect items to have a `key` field, since `getKey()` is customizable

### ISSUE THREAD

**Issue by mshindal** [OfficeDev/office-ui-fabric-react#11484]
### Environment Information

- **Package version(s)**: Latest
- **Browser and OS versions**: N/A

### Please provide a reproduction of the bug in a codepen: 

TypeScript errors don't show up there for some reason

#### Actual behavior:

The `getKey` callback in the `Selection` constructor should allow me to specify my own field to use as a key, but the TypeScript definitions expect a field called `key` to always be there. 

For example, I tried to make it so that `getKey` would return my object's `Id` field. The code looks like this:

```ts
this.selection = new Selection({
      getKey: item => item.Id,
      ...
})
```

I get a TypeScript error that says that: 

> "Property 'Id' does not exist on type 'IObjectWithKey'.'. 

The issue appears to be that `getKey` expects the item passed in to be of type `IObjectWithKey`, which you can see [here](https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/utilities/src/selection/Selection.ts#L9).

This also applies to all of `Selection`'s public methods, like `setItems` (a similar error occurs):

```ts
selection.setItems(items);
```

> Argument of type 'AccountDetails[]' is not assignable to parameter of type 'IObjectWithKey[]'.
  Type 'AccountDetails' has no properties in common with type 'IObjectWithKey'.

#### Expected behavior:

A TypeScript error should not appear in `getKey` when passing an item without a `key` field, or in any of `Selection`'s public methods, like `setItems`. 

### Priorities and help requested:

Are you willing to submit a PR to fix? (Yes, No) Yes

Requested priority: (Blocking, High, Normal, Low) High

Products/sites affected: (if applicable) N/A

**Comment by micahgodbolt** [comment]
`IObjectWithKey` is defined as simply:

```ts
export interface IObjectWithKey {
  key?: string | number;
}
```
So trying to pull out `id` is expected to get an error. 

The key field is option, so you'd want to have a fallback if it is not supplied:

```jsx
  getKey: (item, index) => item.key || index || 'fallback'
```

What exactly are you trying to do here? Is there a scenario where you need access to the selection's ID value?

**Comment by mshindal** [comment]
> What exactly are you trying to do here?

I have a model that has an `Id` field but not a `key` field, and I want to use it with `Selection`. 

So my model looks like:

```ts
interface MyModel {
  Id: string;
  ...
}
```

And I'm trying to tell `Selection` to look at my model's `Id` field when getting a key like so:

```ts
const selection = new Selection({
      // Property 'Id' does not exist on type 'IObjectWithKey'.
      getKey: item => item.Id,
      ...
})
```

And later on I set the selected items like so:

```ts
const myModels: MyModel[] = getMyModels();
//   Type 'MyModel' has no properties in common with type 'IObjectWithKey'.ts(2345)
selection.setItems(myModels);
```

You said:

> trying to pull out id is expected to get an error.

But I don't think that should be an error. If you look at the code for `Selection`, when you pass in a callback for `getKey()`, it [uses the callback you passed in to find the key, instead of looking at the `key` field](https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/utilities/src/selection/Selection.ts#L44). 

So basically what I'm saying is the TypeScript definition is wrong, it should not be limited to `IObjectWithKey` since passing a callback for `getKey()` allows you to use any field as the key, not just the `key` field.

**Comment by ecraig12345** [comment]
I think it would be reasonable to make Selection take a generic item type (defaulting to `IObjectWithKey`). I'll make a PR with that change but probably wait to merge it until @ThomasMichon and others are back in the office and can review.
